subsys/usb/device/class/audio/audio.c

```diff
@@ -541,6 +556,77 @@ static int handle_fu_mute_req(struct usb_audio_dev_data *audio_dev_data,
 	return -EINVAL;
 }
 
+/**
+ * @brief Handler for feature unit volume control requests.
+ *
+ * This function handles feature unit volume control request.
+ *
+ * @param audio_dev_data USB audio device data.
+ * @param setup          Information about the executed request.
+ * @param len		 Size of the buffer.
+ * @param data		 Buffer containing the request result.
+ * @param evt		 Feature Unit Event info.
+ * @param device	 Device part that has been addressed. Applicable for
+ *			 bidirectional device.
+ *
+ * @return 0 if successful, negative errno otherwise.
+ */
+
+static int handle_fu_volume_req(struct usb_audio_dev_data *audio_dev_data,
+				struct usb_setup_packet *setup, int32_t *len, uint8_t **data,
+				struct usb_audio_fu_evt *evt, uint8_t device)
+{
+	uint8_t ch = (setup->wValue) & 0xFF;
+	uint8_t ch_cnt = audio_dev_data->ch_cnt[device];
+	uint8_t *controls = audio_dev_data->controls[device];
+	uint8_t *control_val = &controls[POS(VOLUME, ch, ch_cnt)];
+	int16_t target_vol = 0;
+
+	if (usb_reqtype_is_to_device(setup)) {
+		/* Check if *len has valid value */
+		if (*len != LEN(1, VOLUME)) {
+			LOG_ERR("*len: %d, LEN(1, VOLUME): %d", *len, LEN(1, VOLUME));
+			return -EINVAL;
+		}
+		if (setup->bRequest == USB_AUDIO_SET_CUR) {
+			target_vol = *((int16_t *)*data);
+			if (!IN_RANGE(target_vol, audio_dev_data->volumes.volume_min,
+				      audio_dev_data->volumes.volume_max)) {
+				LOG_ERR("Volume out of range: %d", target_vol);
+				return -EINVAL;
+			}
+			if (target_vol % audio_dev_data->volumes.volume_res != 0) {
+				target_vol = ROUND_UP(target_vol,
+					audio_dev_data->volumes.volume_res);
+			}
+			evt->val = control_val;
+			evt->val_len = *len;
+			*((int16_t *)evt->val) = sys_cpu_to_le16(target_vol);
+			return 0;
+		}
+	} else {
+		if (setup->bRequest == USB_AUDIO_GET_CUR) {
+			*len = LEN(ch_cnt, VOLUME);
+			memcpy(*data, control_val, *len);
+			return 0;
+		} else if (setup->bRequest == USB_AUDIO_GET_MIN) {
+			*len = sizeof(audio_dev_data->volumes.volume_min);
+			memcpy(*data, &(audio_dev_data->volumes.volume_min), *len);
+			return 0;
+		} else if (setup->bRequest == USB_AUDIO_GET_MAX) {
+			*len = sizeof(audio_dev_data->volumes.volume_max);
+			memcpy(*data, &(audio_dev_data->volumes.volume_max), *len);
+			return 0;
+		} else if (setup->bRequest == USB_AUDIO_GET_RES) {
+			*len = sizeof(audio_dev_data->volumes.volume_res);
+			memcpy(*data, &(audio_dev_data->volumes.volume_res), *len);
+			return 0;
+		}
```

Comment:
The memcpy here does not take the fact that USB uses Little-Endian into account. While it will work on LE targets, the code will give incorrect results on BE targets.

Reply:
Thanks, I added Endian convert.